src/dns/query.cpp

```
@@ -3,46 +3,306 @@
-                     cancelled);
+
+    return base;
 }
 
-void Query::cancel(void) {
-    if (cancelled) { // May not be set when we used the default constructor
-        *cancelled = true;
+
+static std::vector<Answer> build_answers_evdns(int code, char type, int count,
+        int ttl, void *addresses) {
+
+    Logger *logger = Logger::global();
+
+    std::vector<Answer> answers;
+
+    if (code != DNS_ERR_NONE) {
+        logger->info("dns - request failed: %d", code);
+        // do not process the results if there was an error
+
+    } else if (type == DNS_PTR) {
+        logger->info("dns - PTR");
+        Answer answer;
+        answer.code = code;
+        answer.ttl = ttl;
+        answer.type = QueryTypeId::PTR;
+        // Note: cast magic copied from libevent regress tests
+        answer.hostname = std::string(*(char **)addresses);
+        logger->info("dns - adding %s", answer.hostname.c_str());
+        answers.push_back(answer);
+    } else if (type == DNS_IPv4_A || type == DNS_IPv6_AAAA) {
+
+        int family;
+        int size;
+        char string[128]; // Is wide enough (max. IPv6 length is 45 chars)
+
+        if (type == DNS_IPv4_A) {
+            family = PF_INET;
+            size = 4;
+            logger->info("dns - IPv4");
+        } else {
+            family = PF_INET6;
+            size = 16;
+            logger->info("dns - IPv6");
+        }
+
+        //
+        // Note: make sure in advance `i * size` won't overflow,
+        // this is here only for robustness.
+        //
+        if (count >= 0 && count <= INT_MAX / size + 1) {
+
+            for (auto i = 0; i < count; ++i) {
+                // Note: address already in network byte order
+                if (inet_ntop(family, (char *)addresses + i * size,
+                              string, sizeof(string)) == NULL) {
+                    logger->warn("dns - unexpected inet_ntop failure");
+                    code = DNS_ERR_UNKNOWN;
```

@hellais setting `code` here as no effect. Since `inet_ntop()` can only fail in corner cases caused by programmer errors, here I'd just throw an exception to exit the program.